server/core/src/main/java/dev/slimevr/tracking/processor/Constraint.kt

```diff
@@ -0,0 +1,183 @@
+package dev.slimevr.tracking.processor
+
+import io.github.axisangles.ktmath.Quaternion
+import io.github.axisangles.ktmath.Vector3
+import kotlin.math.*
+
+/**
+ * Represents a function that applies a rotational constraint.
+ */
+typealias ConstraintFunction = (inputRotation: Quaternion, thisBone: Bone, limit1: Float, limit2: Float, limit3: Float) -> Quaternion
+
+/**
+ * Represents the rotational limits of a Bone relative to its parent,
+ * twist and swing are the max and min when constraintType is a hinge.
+ */
+class Constraint(
+	val constraintType: ConstraintType,
+	twist: Float = 0.0f,
+	swing: Float = 0.0f,
+	allowedDeviation: Float = 0f,
+	maxDeviationFromTracker: Float = 15f,
+) {
+	private val constraintFunction = constraintTypeToFunc(constraintType)
+	private val twistRad = Math.toRadians(twist.toDouble()).toFloat()
+	private val swingRad = Math.toRadians(swing.toDouble()).toFloat()
+	private val allowedDeviationRad = Math.toRadians(allowedDeviation.toDouble()).toFloat()
+	private val maxDeviationFromTrackerRad = Math.toRadians(maxDeviationFromTracker.toDouble()).toFloat()
+	var hasTrackerRotation = false
+
+	/**
+	 * If false solve with minimal movement applied to this link
+	 */
+	var allowModifications = true
+
+	var initialRotation = Quaternion.IDENTITY
+
+	/**
+	 * Apply rotational constraints and if applicable force the rotation
+	 * to be unchanged unless it violates the constraints
+	 */
+	fun applyConstraint(rotation: Quaternion, thisBone: Bone): Quaternion = constraintFunction(rotation, thisBone, swingRad, twistRad, allowedDeviationRad).unit()
+
+	/**
+	 * Force the given rotation to be within allowedDeviation degrees away from
+	 * initialRotation on both the twist and swing axis
+	 */
+	fun constrainToInitialRotation(rotation: Quaternion): Quaternion {
+		val rotationLocal = rotation * initialRotation.inv()
+		var (swingQ, twistQ) = decompose(rotationLocal, Vector3.NEG_Y)
+		swingQ = constrain(swingQ, maxDeviationFromTrackerRad)
+		twistQ = constrain(twistQ, maxDeviationFromTrackerRad)
+		return initialRotation * (swingQ * twistQ)
+	}
+
+	companion object {
+		enum class ConstraintType {
+			TWIST_SWING,
+			HINGE,
+			LOOSE_HINGE,
+			COMPLETE,
+		}
+
+		private fun constraintTypeToFunc(type: ConstraintType) =
+			when (type) {
+				ConstraintType.COMPLETE -> completeConstraint
+				ConstraintType.TWIST_SWING -> twistSwingConstraint
+				ConstraintType.HINGE -> hingeConstraint
+				ConstraintType.LOOSE_HINGE -> looseHingeConstraint
+			}
+
+		private fun decompose(
+			rotation: Quaternion,
+			twistAxis: Vector3,
+		): Pair<Quaternion, Quaternion> {
+			val projection = rotation.project(twistAxis)
+			val twist = Quaternion(sqrt(1.0f - projection.xyz.lenSq()) * if (rotation.w >= 0f) 1f else -1f, projection.xyz).unit()
+			val swing = (rotation * twist.inv()).unit()
+			return Pair(swing, twist)
+		}
+
+		private fun constrain(rotation: Quaternion, angle: Float): Quaternion {
+			val magnitude = sin(angle * 0.5f)
```

Comment:
I would like a comment explaining this if possible, I do not understand why sine is used here or why we multiply by 0.5f. What is the magnitude of?

Reply:
Check the comment I added but the idea is that 0-180 degrees as input to sin(angle * 0.5) maps directly to the length of the vector part of a rotation rotating angle degrees (again assuming a unitized quaternion). So magnitude is the length the vector part of the rotation would have if it was rotating exactly angle degrees and unitized.

Reply:
Sounds good although the comment is cut off with "since a quaternions"

Reply:
fixed, idk how I pulled that one off.